[Confidential-Containers] [2022] Annual Review (#1084)
cncf/toc: reviews/sandbox/2022/2022-Confidential-Containers-annual.md

# Confidential Containers 2022 Annual Review

## Background

Confidential Containers is a CNCF Sandbox project that enables cloud native
confidential computing by taking advantage of a variety of hardware platforms
and technologies. The project was accepted as a CNCF Sandbox in March 2022.

Confidential Containers builds on existing and emerging hardware security

> technologies such as Intel SGX, Intel TDX, AMD SEV and IBM Z Secure Execution,

Reviewer: ```suggestion
technologies such as Intel SGX, Intel TDX, AMD SEV, and IBM Z Secure Execution,
```
I am an Oxford comma kind of guy. Without the comma it reads that AMDSEV and IBM Z Secure Execution requires/expects the use of both technologies.

Author: Fixed, thanks.

> in combination with new software frameworks to help better secure user data in
> use for cloud native workloads.

Reviewer: ```suggestion
in combination with new software frameworks to help better secure data-in-use for users of cloud native workloads.
```

Author: Fixed, thanks.

The main goal of the project is to standardize confidential computing at the

> container level and simplify its consumption in Kubernetes. This is in order to

Reviewer 2: We use the term container or pod level interchangeably. Wondering if `pod level` is more accurate here ?

Author: pod level makes more sense, fixed as well.

Kubernetes users to deploy confidential container workloads using familiar

> workflows and tools without extensive knowledge of underlying confidential

Reviewer: ```suggestion
workflows and tools, without extensive knowledge of underlying confidential
```

Author: Fixed, thanks.

* Contributions are spread over different repositories, showing the width of the
  project:
  * On the client side, the `enclave-cc`, `attestation-agent`, `td-shim` and
    `image-rs` projects take the bulk of the contributions.
  * On the attestation part of the project, most of the contributions went to
    the `kbs` and `attestation-service` repositories.
  * The `cloud-api-adaptor` and `operator` sub-projects merged most of the
    infrastructure related contributions to Confidential Containers.

## Maintainers

The Confidential Containers project maintainership structure is defined as
follows:

* Each sub-project has its own list of maintainers, tracked through both a
  per-repo MAINTAINERS file and a corresponding GitHub team. We have multiple
  instances of individuals maintaining several sub-projects.
* A Technical Steering Committee (TSC) provides decision-making and strategic
  oversight for the project.

The [TSC](https://github.com/confidential-containers/community/blob/main/MAINTAINERS)
is composed of individuals from multiple companies (Alibaba, AMD, IBM, Intel,
Red Hat and Rivos):

- [@ariel-adam](https://github.com/ariel-adam), Ariel Adam, RedHat

> - [@bpradipt](https://github.com/bpradipt), Pradipta Banerjee, IBM

Reviewer 2: ```suggestion
[@ariel-adam](https://github.com/ariel-adam), Ariel Adam, Red Hat
[@bpradipt](https://github.com/bpradipt), Pradipta Banerjee, Red Hat
```

Author: Fixed, sorry.

Reviewer 2: Thanks and np at all

- [@dcmiddle](https://github.com/dcmiddle), Dan Middleton, Intel
- [@fitzthum](https://github.com/fitzthum), Tobin Feldman-Fitzthum, IBM
- [@jiazhang0](https://github.com/jiazhang0), Zhang Jia, Alibaba
- [@jiangliu](https://github.com/jiangliu), Jiang Liu, Alibaba
- [@larrydewey](https://github.com/larrydewey), Larry Dewey, AMD
- [@magowan](https://github.com/magowan), James Magowan, IBM
- [@peterzcst](https://github.com/peterzcst), Peter Zhu, Intel
- [@sameo](https://github.com/sameo), Samuel Ortiz, Rivos

## Adoption

> **Note**
> From CNCF: What do you know about adoption, and how has this changed since your last review / since you joined Sandbox? If you can list companies that are adopters of your project, please do so._

Cloud Providers (Microsoft, Alibaba, IBM) are pusuring offerings which will
include Confidential Containers. They are actively engaged in the community.

> 

Reviewer 3: @sameo can you add the following text? 

Specifically Red Hat now supports confidential with Openshift on Azure (see [setup guide](https://www.redhat.com/en/blog/confidential-containers-azure-openshift-setup-guide)) and Azure support Confidential Containers on AKS as well (see [announcement](https://superuser.openinfra.dev/articles/aligning-with-kata-confidential-containers-to-achieve-zero-trust-operator-deployments-with-aks/)).

Author: > @sameo can you add the following text?
> 
> Specifically Red Hat now supports confidential with Openshift on Azure (see [setup guide](https://www.redhat.com/en/blog/confidential-containers-azure-openshift-setup-guide)) and Azure support Confidential Containers on AKS as well (see [announcement](https://superuser.openinfra.dev/articles/aligning-with-kata-confidential-containers-to-achieve-zero-trust-operator-deployments-with-aks/)).

Slightly reworded, but done.

> 

Reviewer 2: @ariel-adam Can the OpenShift CoCo offering with Azure CVM qualify to be mentioned here ?

Reviewer 3: Yep, I've added it and also regarding AKS.
Can you run the AKS part with Amar?

At recent KubeCon EU, Microsoft invited early adopters to sign up to try out
Confidential Containers in Azure.

## Project Goals

### Performance

> **Note**
> From CNCF: _How has the project performed against its goals since the last review?_

This is the first annual Review for Confidential Containers, so the comparison
baseline is the the moment when Confidential Containers was accepted as a CNCF
Sandbox project, i.e. March 2022.

Since then, the Confidential Containers project published 6
[releases](https://github.com/confidential-containers/documentation/tree/main/releases)
by following a time-based, 6 weeks period release cadence.

> Starting from the initial release, the Confidential Containers delivered on its

Reviewer: ```suggestion
Starting from the initial release, Confidential Containers delivered on its
```

Author: Fixed, thanks.

initial and major goal of enabling seamless use and deployment of unmodified
containers through Kubernetes, across multiple TEEs and hardware platforms.

The project first 6 releases brought the following major key features and
enhancements:

> - A stabilized, enhanced and wider CI that is key to the project's fast release

Reviewer: ```suggestion
- A stabilized, enhanced, and wider CI that is key to the project's fast release
```

Author: Fixed.

  cadence.

> - A hardware backed CI, i.e. a CI that runs test on actual confidential
>   computing hardware instances like e.g. Intel TDX, AMD SEV and Intel SGX.

Reviewer: ```suggestion
- A hardware backed CI, i.e. a CI that runs tests on actual confidential
  computing hardware instances like e.g. Intel TDX, AMD SEV, and Intel SGX.
```

Author: Fixed as well, thanks.

- A Kubernetes Operartor for easily deploying the Confidential Containers

>   runtime, stack and payload on Kubernetes clusters.

Reviewer: ```suggestion
  runtime, stack, and payload on Kubernetes clusters.
```

Author: Fixed, thanks.

- Support for process-based TEEs like e.g. SGX.
- Technical debt reduction, in particular the dependencies on `umoci` and
  `skopeo` are gone.
- A modular, [RATS-compliant](https://datatracker.ietf.org/doc/rfc9334/) Key
  Broker Service (KBS) that implements a full remote attestation service. The
  Confidential Container KBS provides its own attestation service but also
  supports external ones like Intel Amber.

### Current Goals

> **Note**
> From CNCF: _What are the current goals of the project? For example, are you working on major new features? Or are you concentrating on adoption or documentation?_

At a high level, the current project focus is on stability and reliability

> increase, performance improvements and customer adoption expansion.

Reviewer: ```suggestion
increase, performance improvements, and customer adoption expansion.
```

This translates into the following goals:

1. CI/CD improvements
1. Host-Based container images pulling
1. Extended attestation model
1. Policy and secure container configuration
1. Kata containers upstreaming

> 1. Use cases driven implementation

Reviewer 3: @sameo can we change "Use cases driven implementation" to:

Use cases driven implementation replacing feature driven development

## How Can CNCF Help the Project?

## Incubation Criteria

> **Note**
> From CNCF: _Do you think that your project meets the criteria for [incubation](https://github.com/cncf/toc/blob/main/process/graduation_criteria.md#incubating-stage)?_

> We have further work to do to reach the adopter threshold for incubation which we would currently expect to achieve

Reviewer: ```suggestion
We have further work to reach the adopter threshold for incubation, which we would currently expect to achieve
```

Reviewer 3: I'd change it to the following:

We are in the process of getting direct adopters to consume the Confidential Container solution both members of our community and external. Our target is to have at least 3 direct adopters by the end of 2023 helping us to grow to the next level of incubation.

Reviewer: Good wording, @ariel-adam. Just added some punctuation.

```suggestion
We are in the process of getting direct adopters to consume the Confidential Container solution; both members of our community and external. Our target is to have at least 3 direct adopters by the end of 2023, helping us to grow to the next level of incubation.
```

Author: I took @larrydewey suggestion, thanks.
